src/validation.cpp

```diff
@@ -3834,7 +3834,7 @@ CBlockIndex * CChainState::InsertBlockIndex(const uint256& hash)
 
 bool CChainState::LoadBlockIndex(const Consensus::Params& consensus_params, CBlockTreeDB& blocktree)
 {
-    if (!blocktree.LoadBlockIndexGuts(consensus_params, [this](const uint256& hash) EXCLUSIVE_LOCKS_REQUIRED(cs_main) { return this->InsertBlockIndex(hash); }))
+    if (!blocktree.LoadBlockIndexGuts(consensus_params, [this](const uint256& hash) { AssertLockHeld(cs_main); return this->InsertBlockIndex(hash); }))
```

Comment:
I think even if it compiles, the compiler couldn't enforce the lock annotations for lambdas anyway, since they could be used across compilation units?

Reply:
I'm not sure the implication, but I think this is good - `EXCLUSIVE_LOCKS_REQUIRED(cs_main)` is also enforced via the `CChainState::LoadBlockIndex` declaration, and `AssertLockHeld` looks into `g_lockstack`, which is built up in the usual way. We could perhaps do without `AssertLockHeld` given the `LoadBlockIndex` declaration. Thoughts?

Reply:
Removing as it's also asserted in `InsertBlockIndex`

Reply:
Scratch that, a thread-safety-warning is issued absent this assert, thanks to the `EXCLUSIVE_LOCKS_REQUIRED` on `InsertBlockIndex`.